File: ea/generate_combination.py
from pprint import pprint


def helper(s, i, path, res):
  if i == len(s):
    res.append(sorted(path))
    return

  helper(s, i+1, path+[s[i]], res)
  for j in range(len(path)):
    helper(s, i+1, path[:j]+[path[j]+s[i]]+path[j+1:], res)


def solve(s):
  res = list()
  helper(s, 1, [s[0]], res)
  return sorted(res)


pprint(solve('DDFM'))

File: ea/generate_combination.js
const _ = require('lodash');

function generateCombinations(sidePosition) {
  const helper = (s, i, path, res) => {
    if (i == s.length) return res.push(_.sortBy(path));

    helper(s, i+1, _.concat(path, s[i]), res);
    for (const j of _.range(0, path.length)) {
      helper(s, i+1, _.concat(_.slice(path, 0, j), path[j]+s[i], ..._.slice(path, j+1)), res);
    }
  };

  res = new Set();
  helper(sidePosition, 1, [sidePosition[0]], res)
  return _.sortBy(res);
}

console.log(generateCombinations('DDFM'));